Create cryptogram validation

LaurenHimbeault opened the pull request:
Merge these?

Commit:
Changed the arrangement of the Puzzle class because CODING STANDARDS said that the constructor should be at the top

Commit:
Basic Creation of class Cryptogram extending puzzle

Commit:
Altered more of the structure

in Puzzle and IPlayable I changed the name of the boolean method getIsCompleted to just isCompleted to follow CODING STANDARDS that we wrote :)

Added some basic validation to the isCompleted method in Cryptogram

Commit:
Added a bunch of logic and tests around the Cryptogram class

You can now:

create a cryptogram and it will generate random ciphertext for you
print it it out as well (see tests)
I think I covered lots of cases but if people see more cases, please help me cover them

Commit:
See previous commit message(this shouldve all been together, about the testing and stuffz)

Commit:
Wrote some more tests

willvoutier commented on pg13/models/Cryptogram.java:
```diff
@@ -0,0 +1,202 @@
+	 *  @param String author - whoever designed the cryptogram
+	 *  @param String title - name of cryptogram
+	 *  @param Date dateCreated - date of creation of puzzle
+	 *  @param String pt - plaintext given by the user for the puzzle
```
Can we try to use more meaningful names for variable names, particularly in constructors? "pt" doesn't convey any meaning on it's own and requires the Java doc to explain its semantic value.

willvoutier commented on pg13/models/Cryptogram.java:
```diff
@@ -0,0 +1,202 @@
+	private ArrayList<Character> generateAlphabet() {
+		ArrayList <Character> key =  new ArrayList<Character> ();
+		
+		key.add('A');
```
I feel like we can do better than this. Loop?

willvoutier commented on tests/models/CryptogramTest.java:
```diff
@@ -0,0 +1,116 @@
+		{
+			Cryptogram cryptogram;
+
+			System.out.println("\nStarting testCryptogram");
```
We shouldn't have console output in our tests. JUnit, being a testing framework, handles the logic of setting up, tearing down, and notifying us about failed tests.

willvoutier commented on tests/models/CryptogramTest.java:
```diff
@@ -0,0 +1,116 @@
+			System.out.println("Finished testCryptogramGeneralData");
+		}
+
+		public void testCipherTextWorks1()
```
When naming a test, you want to ask yourself what _specifically_, you are testing. The fact that you're testing that something "works" is innate in the fact that you're testing something at all. If this test ever fails for whatever reason, the test name should provide a basic orientation for where something went wrong and where to fix it.

Does that makes sense?

willvoutier commented on tests/models/CryptogramTest.java:
```diff
@@ -0,0 +1,116 @@
+			System.out.println("\nStarting testCipherTextWorks1");		
+			cryptogram = new Cryptogram("Lauren Slusky", "Practice Cryptogram", new Date(), "This is a test");
+			assertEquals("This is a test", cryptogram.getPlainText());
+			assertNotSame("This is a test", cryptogram.getCipherText());
```
Is this enforced anywhere, or will this fail if we unluckily get random selections in the algorithm that match up all characters to themselves?

willvoutier commented on tests/models/PuzzleTests.java:
```diff
@@ -0,0 +1,17 @@
+    public static Test suite()
+    {
+        suite = new TestSuite("Puzzle tests");
+      //  suite.addTestSuite(CalculateGPATest.class);
```
We probably don't want this in our code.

willvoutier commented on tests/models/PuzzleTests.java:
```diff
@@ -0,0 +1,17 @@
+      //  suite.addTestSuite(CalculateGPA2Test.class);
+       return suite;
+    }
+}
```
It's good practice to have an empty line at the end of every text file. Because [stupid stuff](http://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file).

LaurenHimbeault replied:
1) printing to the console in the tests:I did it because there is no other
way to verify that the encryption is happening correctly since it's random.
This can be removed if we don't want to show we tested that. Let me know

2) the extra line can be added bit please add it to the coding standards

3)array list: I don't understand your question Tbh, what is the issue with
generate keys?
On 2013-05-26 4:51 PM, "williamhc" notifications@github.com wrote:

> In tests/models/PuzzleTests.java:
> 
> > +
> > +import junit.framework.Test;
> > +import junit.framework.TestSuite;
> > +
> > +public class PuzzleTests
> > +{
> > -   public static TestSuite suite;
> >   +
> > -    public static Test suite()
> > -    {
> > -        suite = new TestSuite("Puzzle tests");
> > -      //  suite.addTestSuite(CalculateGPATest.class);
> > -      //  suite.addTestSuite(CalculateGPA2Test.class);
> > -       return suite;
> > -    }
> >   +}
> 
> It's good practice to have an empty line at the end of every text file.
> Because stupid stuffhttp://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/williamhc/PG13/pull/5/files#r4395226
> .

Commit:
Added some logic to make sure no letter maps on to itself

fixed the test suites so that I dont print to console but instead to an assert and call a decrypt method to check if the encryption worked properly

added a space as per Will's request regarding puzzletests an the end of the file :)

altered some variable names to make them easier to use and whatnot

willvoutier commented:
Looks great. I'll add the empty line at end of file to the coding standards. Merging.
